### ISSUE TITLE
ip6tables causes daemon to crash (experimental)

### ISSUE THREAD

**Issue by bboehmke** [moby/moby#41774]
**Description**

With enabled ip6tables (including the experimental option) the daemon crashes on startup.

**Steps to reproduce the issue:**

Start the daemon with:
```
dockerd --ipv6 --fixed-cidr-v6 fd00:dead:beef::/48 --ip6tables --experimental
```

**Describe the results you received:**
```
INFO[2020-12-10T15:23:18.538460939+01:00] Starting up                                  
WARN[2020-12-10T15:23:18.538675516+01:00] Running experimental build                   
INFO[2020-12-10T15:23:18.539879145+01:00] parsed scheme: "unix"                         module=grpc
INFO[2020-12-10T15:23:18.540005907+01:00] scheme "unix" not registered, fallback to default scheme  module=grpc
INFO[2020-12-10T15:23:18.540109662+01:00] ccResolverWrapper: sending update to cc: {[{unix:///run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}  module=grpc
INFO[2020-12-10T15:23:18.540210588+01:00] ClientConn switching balancer to "pick_first"  module=grpc
INFO[2020-12-10T15:23:18.542245104+01:00] parsed scheme: "unix"                         module=grpc
INFO[2020-12-10T15:23:18.542271637+01:00] scheme "unix" not registered, fallback to default scheme  module=grpc
INFO[2020-12-10T15:23:18.542293888+01:00] ccResolverWrapper: sending update to cc: {[{unix:///run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}  module=grpc
INFO[2020-12-10T15:23:18.542363497+01:00] ClientConn switching balancer to "pick_first"  module=grpc
INFO[2020-12-10T15:23:18.556494462+01:00] [graphdriver] using prior storage driver: overlay2 
WARN[2020-12-10T15:23:18.561225739+01:00] Your kernel does not support swap memory limit 
WARN[2020-12-10T15:23:18.564152713+01:00] Your kernel does not support CPU realtime scheduler 
INFO[2020-12-10T15:23:18.564628857+01:00] Loading containers: start.                   
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5590ba437939]

goroutine 1 [running]:
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.(*bridgeNetwork).setupIP6Tables(0xc0003a9260, 0xc000ad7790, 0xc0003a9140, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/setup_ip_tables.go:128 +0x79
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.(*bridgeSetup).apply(0xc0009b4750, 0xc00008e180, 0x8)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/setup.go:17 +0x7c
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.(*driver).createNetwork(0xc000a82200, 0xc000ad7790, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/bridge.go:809 +0x7e7
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.(*driver).populateNetworks(0xc000a82200, 0x5, 0x5590bb13bf7a)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/bridge_store.go:62 +0x28d
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.(*driver).initStore(0xc000a82200, 0xc000a85b90, 0x0, 0xc00083a8c0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/bridge_store.go:35 +0x211
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.(*driver).configure(0xc000a82200, 0xc000a85b90, 0x5590bbe3dae0, 0xc000a858c0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/bridge.go:439 +0x20d
github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge.Init(0x5590bc312760, 0xc0006a6d80, 0xc000a85b90, 0xc000a85b90, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drivers/bridge/bridge.go:169 +0x9c
github.com/docker/docker/vendor/github.com/docker/libnetwork/drvregistry.(*DrvRegistry).AddDriver(...)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/drvregistry/drvregistry.go:72
github.com/docker/docker/vendor/github.com/docker/libnetwork.New(0xc000a82180, 0x9, 0x10, 0xc0005015f0, 0xc000a85860, 0xc000a82180, 0x9)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:221 +0x4f0
github.com/docker/docker/daemon.(*Daemon).initNetworkController(0xc00000c1e0, 0xc0000c5600, 0xc000a85860, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/daemon_unix.go:854 +0xa9
github.com/docker/docker/daemon.(*Daemon).restore(0xc00000c1e0, 0xc000713900, 0xc00015c380)
	/go/src/github.com/docker/docker/daemon/daemon.go:467 +0x50b
github.com/docker/docker/daemon.NewDaemon(0x5590bc341a20, 0xc000713900, 0xc0000c5600, 0xc0005015f0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/daemon.go:1117 +0x2be7
main.(*DaemonCli).start(0xc000501350, 0xc00084b920, 0x0, 0x0)
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:195 +0x743
main.runDaemon(...)
	/go/src/github.com/docker/docker/cmd/dockerd/docker_unix.go:13
main.newDaemonCommand.func1(0xc0006fadc0, 0xc000504500, 0x0, 0x5, 0x0, 0x0)
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:34 +0x7c
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc0006fadc0, 0xc00004e1f0, 0x5, 0x5, 0xc0006fadc0, 0xc00004e1f0)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:850 +0x462
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0006fadc0, 0x0, 0x0, 0x10)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:958 +0x34b
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:895
main.main()
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:97 +0x191
```

**Describe the results you expected:**

Startup without exception

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.0
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        7287ab3
 Built:             Tue Dec  8 18:59:49 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.0
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       eeddea2
  Built:            Tue Dec  8 18:57:58 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.4.2-docker)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 20.10.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.19.0-13-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 1.902GiB
 Name: docker-test
 ID: J7KW:ISZL:54TP:KIGF:IFYC:YTTF:635N:LD2C:XMRZ:NTRG:4RJD:IXA5
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Hetzner cloud CX11 VM

**Comment by thaJeztah** [comment]
Thanks for reporting

- relates to https://github.com/moby/moby/pull/41622 IPv6 iptables config option 
- relates to https://github.com/moby/libnetwork/pull/2572 Enable IPv6 NAT
    vendored in https://github.com/moby/moby/pull/41604


/cc @arkodg @bboehmke

**Comment by bboehmke** [comment]
I think I found the root cause of this issue. If docker is installed the daemon is also started which creates an interface without an IPv6 address. If now the IPv6 is enabled the IPv6 address is not added to the existing interface which causes the daemon to crash on startup because the forward address is not set for IPv6 (`nil`).

If the interface is removed and the daemon is restarted it starts also with ip6tables enabled.
(After that we run into the next issue but here I have to investigate more)

I am not sure if this is by design or a bug but in combination with `--ip6tables` it is not so ideal.

**Comment by thaJeztah** [comment]
Panic is happening here; https://github.com/moby/moby/blob/v20.10.0/vendor/github.com/docker/libnetwork/drivers/bridge/setup_ip_tables.go#L127-L130

```go
maskedAddrv6 := &net.IPNet{
	IP:   i.bridgeIPv6.IP.Mask(i.bridgeIPv6.Mask),
	Mask: i.bridgeIPv6.Mask,
}
```

So either `i` (`*bridgeInterface`) or `i.bridgeIPv6` (`*net.IPNet`) could be `nil`. 

The latter one is  set in two places (at a quick glance) inside `setupBridgeIPv6(()` https://github.com/moby/moby/blob/v20.10.0/vendor/github.com/docker/libnetwork/drivers/bridge/setup_ipv6.go#L45-L47

```go
// Store bridge network and default gateway
i.bridgeIPv6 = bridgeIPv6
i.gatewayIPv6 = i.bridgeIPv6.IP
```

And later overwritten; https://github.com/moby/moby/blob/v20.10.0/vendor/github.com/docker/libnetwork/drivers/bridge/setup_ipv6.go#L57-L59

```go
// Store the user specified bridge network and network gateway and program it
i.bridgeIPv6 = config.AddressIPv6
i.gatewayIPv6 = config.AddressIPv6.
```

(I think I actually had a branch somewhere that made some changes in the same area for IPv4)

That `bridgeIPv6` is also a global variable set on `init()`; https://github.com/moby/moby/blob/v20.10.0/vendor/github.com/docker/libnetwork/drivers/bridge/setup_ipv6.go#L23-L30
Although it should panic if it fails to be set (had some changes locally to remove the `init()` https://github.com/moby/libnetwork/pull/2599), but don't think the problem should be there 🤔

**Comment by thaJeztah** [comment]
> I think I found the root cause of this issue. If docker is installed the daemon is also started which creates an interface without an IPv6 address. If now the IPv6 is enabled the IPv6 address is not added to the existing interface which causes the daemon to crash on startup because the forward address is not set for IPv6

oh! your comment didn't show up yet; thanks, that's useful information

**Comment by thaJeztah** [comment]
Just now realised I pinged you in my first comment, not seeing it was you who reported the issue

**Comment by bboehmke** [comment]
The daemon currently reacts really crazy to the ip6tables option. 
I am 100% that I get it working in the past but I have no clue why it is making so much trouble now.

**Comment by bboehmke** [comment]
So i found now multiple issues that I have not seen before:

1. Daemon crashes if the `docker0` interface does not have an IPv6 address (as described above)
2. On interface creation the IPv6 address is (some times) added after the `setupIP6Tables` (which causes 1.)
3. The IPv6 route is also added before the interface is up: (same for user networks)
```
DEBU[2020-12-10T16:25:15.393647574+01:00] Did not find any interface with name docker0: Link not found 
DEBU[2020-12-10T16:25:15.394025442+01:00] Setting bridge mac address to 02:42:6c:5d:e0:3f 
DEBU[2020-12-10T16:25:15.398282827+01:00] Assigning address to bridge interface docker0: 172.17.0.1/16 
DEBU[2020-12-10T16:25:15.398848139+01:00] Adding route to IPv6 network fd00:dead:beef::1/48 via device docker0 
ERRO[2020-12-10T16:25:15.399066140+01:00] Could not add route to IPv6 network fd00:dead:beef::1/48 via device docker0: network is down 
```
4. There are now `ip6tables` call to IPv4 address (the opposite of #41685)
```
DEBU[2020-12-10T16:26:18.729635029+01:00] /usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p tcp -d 0/0 --dport 80 -j DNAT --to-destination 172.17.0.2:80 ! -i docker0] 
DEBU[2020-12-10T16:26:18.731600667+01:00] /usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p tcp -d 0/0 --dport 80 -j DNAT --to-destination 172.17.0.2:80 ! -i docker0] 
WARN[2020-12-10T16:26:18.737597966+01:00] Failed to allocate and map port 80-80:  (iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 80 -j DNAT --to-destination 172.17.0.2:80 ! -i docker0: ip6tables v1.8.2 (nf_tables): Bad IP address "172.17.0.2"

Try `ip6tables -h' or 'ip6tables --help' for more information.
 (exit status 2)) 
```
(here also the log output is wrong `iptables --wait ...` should be `ip6tables --wait `)


I will try to dig a bit deeper but currently I get a bit overwhelmed by the amount and kind of issues.

**Comment by arkodg** [comment]
https://github.com/moby/libnetwork/pull/2600 should solve the panic

Will need to find a dedicated cycle to look into the Iptables version mismatch issue

**Comment by bboehmke** [comment]
1. The change from @arkodg seems to fix this

2. I was not able to reproduce this anymore

3. Seems not to be a bigger issue but needs more investigation

4. I found a fix for this in `vendor/github.com/docker/libnetwork/portmapper/mapper_linux.go`: the `checkIP()` function is still not working correctly. I have already a fix for this and will provide an PR soon but sadly we now run into a new issue (see 5.)

5. The userland proxy (dummy and real one) causes a `bind: address already in use` because the default bind address `0.0.0.0` seems to also bind to IPv6 which causes the second call for IPv6 to fail. Here I am still working on a solution.

**Comment by bboehmke** [comment]
After some hours of debugging I found the real cause of 4 (see https://github.com/moby/libnetwork/pull/2602/commits/9c68849e2b1c53db5ed862320893274abd366907).

I also was able to solve the problems with the userland proxy.

For all fixes see moby/libnetwork#2602

**Comment by bboehmke** [comment]
tested with release 20.10.2 and everything seems to work now
